File: packages/indexer/scripts/populate.ts
import 'dotenv/config';
import { Contract } from '@prisma/client';
import prisma from '../src/prisma';

const CONTRACTS: Pick<
  Contract,
  | 'name'
  | 'type'
  | 'address'
  | 'chain'
  | 'symbol'
  | 'deployedBlock'
  | 'targetGroups'
>[] = [
  {
    address: '0xa9d54f37ebb99f83b603cc95fc1a5f3907aaccfd',
    name: 'Pikaboss',
    symbol: 'pika',
    deployedBlock: BigInt(16628745),
    targetGroups: ['earlyHolder', 'whale'],
    chain: 'Ethereum',
    type: 'ERC20',
  },
  {
    address: '0x02e7f808990638e9e67e1f00313037ede2362361',
    name: 'KiboShib',
    symbol: 'KIBSHI',
    deployedBlock: BigInt(16140853),
    targetGroups: ['earlyHolder', 'whale'],
    chain: 'Ethereum',
    type: 'ERC20',
  },
];

const DEV_FIDS: number[] = [54, 12783, 20559];

const { IS_PULL_REQUEST, NODE_ENV } = process.env;

const populate = async () => {
  // Run this script only in PRs (i.e. Render preview environments)
  // and non-production environments
  if (IS_PULL_REQUEST === 'true' || NODE_ENV !== 'production') {
    for (const contract of CONTRACTS) {
      await prisma.contract.upsert({
        update: contract,
        create: contract,
        where: {
          address_chain: {
            address: contract.address,
            chain: contract.chain,
          },
        },
      });
    }

    // Index the dev Merkle trees.
    // We just need to import the file to trigger the indexing
    require('../src/indexMerkleTree');

    // Get one of the created Merkle tree
    const devTree = await prisma.merkleTree.findFirst({
      select: {
        id: true,
      },
    });

    if (!devTree) {
      throw new Error('No Merkle tree found!');
    }

    // Create some dummy FID attestations
    await prisma.fidAttestation.createMany({
      data: DEV_FIDS.map(fid => ({
        fid,
        signInSig: Buffer.from('dummy'),
        attestation: Buffer.from('dummy'),
        treeId: devTree.id,
      })),
      skipDuplicates: true,
    });
  }
};

populate();

File: packages/indexer-rs/src/bin/populate.rs
use indexer_rs::{
    contract::{upsert_contract, Contract, ContractType},
    contracts::{erc20::erc20_contracts, erc721::erc721_contracts},
    eth_rpc::Chain,
    postgres::init_postgres,
    utils::{dotenv_config, is_prod},
    GroupType,
};

#[tokio::main]
async fn main() {
    dotenv_config();

    // Connect to the database.
    let client = init_postgres().await;

    if is_prod() {
        // Populate all contracts in production.

        let mut contracts = vec![];

        contracts.extend(erc721_contracts().iter().cloned());
        contracts.extend(erc20_contracts().iter().cloned());

        for contract in contracts {
            upsert_contract(&client, &contract).await.unwrap();
        }
    } else {
        // Only populate few small contracts in development and preview environments.
        let contracts = [
            Contract {
                id: 0,
                address: "0xa9d54f37ebb99f83b603cc95fc1a5f3907aaccfd".to_string(),
                chain: Chain::Mainnet,
                symbol: "pika".to_string(),
                name: "Pikaboss".to_string(),
                deployed_block: 16628745,
                target_groups: vec![GroupType::EarlyHolder, GroupType::Whale],
                contract_type: ContractType::ERC20,
            },
            Contract {
                id: 0,
                address: "0x02e7f808990638e9e67e1f00313037ede2362361".to_string(),
                chain: Chain::Mainnet,
                symbol: "KIBSHI".to_string(),
                name: "KiboShib".to_string(),
                deployed_block: 16140853,
                target_groups: vec![GroupType::EarlyHolder, GroupType::Whale],
                contract_type: ContractType::ERC20,
            },
            Contract {
                id: 0,
                address: "0x5Af0D9827E0c53E4799BB226655A1de152A425a5".to_string(),
                symbol: "MIL".to_string(),
                name: "Milady".to_string(),
                chain: Chain::Mainnet,
                deployed_block: 13090020,
                target_groups: vec![GroupType::AllHolders],
                contract_type: ContractType::ERC721,
            },
        ];

        for contract in &contracts {
            upsert_contract(&client, contract).await.unwrap();
        }
    }
}
